proposals/text/3784-input-output-shaper.md

```diff
@@ -0,0 +1,247 @@
+              type: InputOutputShaper
+            pipelines:
+              - name: query
+                nodes:
+                  - name: shaper
+                    inputs:
+                      - Query
+
+```
+
+We need the "default parameter" functionality to support YAML files definitions that are more concise and less
+error-prone.
+
+
+### Order of function invocation
+
+The order of function invocation is important. The functions are invoked in the order they are defined in the YAML.
+In the example below, we have two input variables: `query` and `documents`. The `query` variable is expanded to a
+list of strings and stored in the variable `questions`. The `documents` variable is then contracted and the
+variable `questions` is immediately used as the `num_tokens` keyword parameter to the `contract_docs` function.
+
+```yaml
+            components:
+            - name: shaper
+              params:
+                inputs:
+                    query:
+                      func: expand
+                      output: questions
+                      params:
+                        expand_target: query
+                        size:
+                          func: len
+                          params:
+                            - documents
+                    documents:
+                      func: contract_docs
+                      output: documents
+                      params:
+                        docs: documents
+                        delimiter: " "
+                        num_tokens:
+                            func: len
+                            params:
+                                - questions
+              type: InputOutputShaper
+            pipelines:
+              - name: query
+                nodes:
+                  - name: shaper
+                    inputs:
+                      - Query
+```
+
+### Security
+
+The InputOutputShaper component prevents arbitrary code execution. Only the functions we have specified in the
+InputOutputShaper component are allowed to be invoked. The functions are specified in the constructor using immutable
+data structures.
+
+This security measure might be relaxed in the future to allow users to specify their own functions. However, this
+change will require a more thorough security review.
+
+# Drawbacks
+
+Although a "normal use" of PromptNodes would not trigger a need for InputOutputShaper there are cases where its
+use is necessary. In cases where we can only use pipeline definitions to configure the pipeline, we need to use it.
+
+
+- Implementation cost is not high. The InputOutputShaper component is already implemented and tested.
+- InputOutputShaper can't be defined outside of Haystack.
+- InputOutputShaper could turn out to be useful in other use cases as well - i.e. declarative pre/post processing.
+- InputOutputShaper integrates well with PromptNodes and other components.
+- No braking changes to existing components.
+
```

Comment:
Most of these points are not drawbacks :)

Reply:
Thank @mkkuemmel - I answered questions in the proposal template mindlessly. Noted.